### PR TITLE
lochness pull for Mediaflux

### PR DESCRIPTION
**Pull request by tashrifbillah**

### CODE REVIEW

File: lochness/mediaflux/__init__.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+            break
+        fo.write(buf)
+
+    fo.flush()
+    os.fsync(fo.fileno())
+    fo.close()
+    return fo.name
+
+
+def verify(f, content_hash, key=None, compress=False):
```
Comment (tashrifbillah) [inline lochness/mediaflux/__init__.py:129]: To be removed ...

File: lochness/mediaflux/__init__.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    # verify the file and rename to final local destination
+    logger.debug(f'verifying temporary file {tmp_name}')
+    verify(tmp_name, box_file_object.sha1, key=key, compress=compress)
+    os.chmod(tmp_name, 0o0644)
+    os.rename(tmp_name, local_fullfile)
+
+
+
+
+def _savetemp(content, dirname=None, compress=False):
```
Comment (tashrifbillah) [inline lochness/mediaflux/__init__.py:109]: To be removed ...

File: lochness/mediaflux/__init__.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+                logger.debug(f'deleting file on box {box_fullpath}')
+                _delete(box_file_object, box_fullpath)
+        except HashRetryError:
+            msg = f'too many failed attempts downloading {box_fullpath}'
+            raise DownloadError(msg)
+
+
+
+
+def _save(box_file_object, box_fullpath, local_fullfile, key, compress):
```
Comment (tashrifbillah) [inline lochness/mediaflux/__init__.py:78]: To be removed ...

File: lochness/mediaflux/__init__.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+def base(Lochness, study_name):
+    ''' get module-specific base box diretory '''
+    return Lochness.get('mediaflux', {}) \
+                   .get(study_name, {}) \
+                   .get('namespace', '')
+
+
+
+
+def save(box_file_object: boxsdk.object.file,
```
Comment (tashrifbillah) [inline lochness/mediaflux/__init__.py:42]: To be removed ...

**Review by tashrifbillah — COMMENTED** [review]
TODO:
- [ ] encryption feature after download
- [ ] `sync.py` instruction in mediaflux.md
- [ ] mediaflux.md-->mediaflux.rst
- [ ] put pieces of info in the combined doc
- [ ] remove dead codes from `lochness/mediaflux/__init__.py`

File: lochness/keyring/__init__.py [hunk 5]
```diff
@@ -24,6 +26,79 @@ def box_api_token(Lochness, key):
...
+        raise KeyringError('\'{0}\' not in keyring'.format(key))
+
+    # allow having both TOKEN and PASSWORD in keyring, prioritize TOKEN
+    if 'TOKEN' in Lochness['keyring'][key]:
+        credentials = ['HOST', 'PORT', 'TRANSPORT', 'TOKEN']
+    else:
+        credentials = ['HOST', 'PORT', 'TRANSPORT', 'TOKEN', 'DOMAIN', 'USER', 'PASSWORD']
+
+    auths= Lochness['keyring'][key]
+    mflux_cfg= os.path.join(Lochness['phoenix_root'], 'mflux.cfg')
```
Comment (kcho) [inline lochness/keyring/__init__.py:88]: Is there a specific reason why you are creating an extra file tha thas the credentials?

Comment (tashrifbillah) [reply lochness/keyring/__init__.py:88]: Yes, there is a specific reason. Mediaflux command line client would only accept a file to read login credentials from. There is no way we can provide them through the command line.

**Review by kcho — APPROVED** [review]
I assume Mediaflux doesn't have python tools to use the APIs ? 

All looks great to me. Especially the docs.

**Comment by tashrifbillah** [comment]
> I assume Mediaflux doesn't have python tools to use the APIs ?

There is one but not as affluent as the command line one. The latter has a few additional features including checksum checking which is very useful for us.

**Comment by kcho** [comment]
> > I assume Mediaflux doesn't have python tools to use the APIs ?
> 
> There is one but not as affluent as the command line one. The latter has a few additional features including checksum checking which is very useful for us.

How would you go about installing the `CLI` on the user side? Is it simple enough to install the `CLI` for the users?

**Comment by tashrifbillah** [comment]
Just unzipping, yes.

File: docs/source/quick_start.rst [hunk 6]
```diff
@@ -31,19 +31,29 @@ the *keyring*. Here's an example of a decrypted keyring file ::
         "URL": "https://beiwe.example.org",
         "ACCESS_KEY": "...",
         "SECRET_KEY": "..."
-      }
+      },
```
Comment (tashrifbillah) [inline docs/source/quick_start.rst:34]: This comma was missing. I have inserted that @kcho

File: docs/source/quick_start.rst [hunk 7]
```diff
@@ -31,19 +31,29 @@ the *keyring*. Here's an example of a decrypted keyring file ::
...
         "PASSWORD": "..."
       },
 
-      "box.xxxxx": {
+      "box.example": {
         "CLIENT_ID": "...",
         "CLIENT_SECRET": "...",
         "API_TOKEN": "..."
-        }
+      },
```
Comment (tashrifbillah) [inline docs/source/quick_start.rst:46]: This comma was missing. I have inserted that @kcho

File: docs/source/quick_start.rst [hunk 8]
```diff
@@ -31,19 +31,29 @@ the *keyring*. Here's an example of a decrypted keyring file ::
...
 
       "xnat.example": {
         "URL": "https://chpe-xnat.example.harvard.edu",
         "USERNAME": "...",
         "PASSWORD": "..."
       },
 
-      "box.xxxxx": {
+      "box.example": {
```
Comment (tashrifbillah) [inline docs/source/quick_start.rst:42]: I have made the identifier the same as the rest of the sub sections. @kcho

File: docs/source/configuration_file.rst [hunk 9]
```diff
@@ -146,6 +146,11 @@ The resulting section should look as follows ::
                          pattern: '.*\.mov'
```
Comment (kcho) [inline docs/source/configuration_file.rst:146]: would you change this line to `pattern: '*.mov'` please?

Comment (tashrifbillah) [reply docs/source/configuration_file.rst:146]: Would the change work in the current setting? I haven't implemented my template update on Box yet.

Comment (tashrifbillah) [reply docs/source/configuration_file.rst:146]: Will be done later ...

**Pull request merged**